### ISSUE TITLE
library not working for Arduino Nano Every board

### ISSUE THREAD

**Issue by tdenslow** [AKJ7/TM1637#10]
Error: Expected identifier before numeric constant   
 log included below

/home/builder/opt/arduino-builder/arduino-builder -compile -core-api-version 10611 -hardware /home/builder/opt/arduino-builder/hardware -hardware /home/builder/.arduino15/packages -tools /home/builder/opt/arduino-builder/tools -tools /home/builder/.arduino15/packages -built-in-libraries /home/builder/opt/libraries/latest -logger humantags -fqbn arduino:megaavr:nona4809:mode=on -build-cache /tmp -build-path /tmp/742609487/build -verbose -prefs runtime.tools.arduinoOTA.path=/home/builder/.arduino15/packages/arduino/tools/arduinoOTA/1.3.0 -prefs runtime.tools.avr-gcc.path=/home/builder/.arduino15/packages/arduino/tools/avr-gcc/7.3.0-atmel3.6.1-arduino5 -prefs runtime.tools.avrdude.path=/home/builder/.arduino15/packages/arduino/tools/avrdude/6.3.0-arduino17 -prefs runtime.tools.bossac.path=/home/builder/.arduino15/packages/industruino/tools/bossac/1.7.0-industruino -prefs runtime.tools.openocd.path=/home/builder/.arduino15/packages/Intel/tools/openocd/0.9.0+0.1 -prefs runtime.tools.flashpack.path=/home/builder/.arduino15/packages/Intel/tools/flashpack/2.0.0 -prefs runtime.tools.ctags.path=/home/builder/.arduino15/packages/builtin/tools/ctags/5.8-arduino11 -prefs runtime.tools.arm-linux-gcc.path=/home/builder/.arduino15/packages/arduino/tools/arm-linux-gcc/4.9.3 -prefs runtime.tools.arm-none-eabi-gcc.path=/home/builder/.arduino15/packages/arduino/tools/arm-none-eabi-gcc/7-2017q4 -prefs runtime.tools.nrf5x-cl-tools.path=/home/builder/.arduino15/packages/arduino/tools/nrf5x-cl-tools/9.3.1 -prefs runtime.tools.linuxuploader.path=/home/builder/.arduino15/packages/arduino/tools/linuxuploader/1.5.1 -prefs runtime.tools.CMSIS.path=/home/builder/.arduino15/packages/arduino/tools/CMSIS/4.5.0 -prefs runtime.tools.esptool.path=/home/builder/.arduino15/packages/esp8266/tools/esptool/2.5.0-3-20ed2b9 -prefs runtime.tools.mkspiffs.path=/home/builder/.arduino15/packages/esp8266/tools/mkspiffs/2.5.0-3-20ed2b9 -prefs runtime.tools.arduino101load.path=/home/builder/.arduino15/packages/Intel/tools/arduino101load/2.0.1 -prefs runtime.tools.core2-32-poky-linux.path=/home/builder/.arduino15/packages/Intel/tools/core2-32-poky-linux/1.6.2+1.0 -prefs runtime.tools.i586-poky-linux-uclibc.path=/home/builder/.arduino15/packages/Intel/tools/i586-poky-linux-uclibc/1.6.2+1.0 -prefs runtime.tools.CMSIS-Atmel.path=/home/builder/.arduino15/packages/arduino/tools/CMSIS-Atmel/1.2.0 -prefs runtime.tools.xtensa-lx106-elf-gcc.path=/home/builder/.arduino15/packages/esp8266/tools/xtensa-lx106-elf-gcc/2.5.0-3-20ed2b9 -prefs runtime.tools.arc-elf32.path=/home/builder/.arduino15/packages/Intel/tools/arc-elf32/1.6.9+1.0.1 -prefs runtime.tools.sketchUploader.path=/home/builder/.arduino15/packages/Intel/tools/sketchUploader/1.6.2+1.0 -prefs runtime.tools.dfu-util.path=/home/builder/.arduino15/packages/arduino/tools/dfu-util/0.9.0-arduino2 -prefs runtime.tools.x86-linux-gcc.path=/home/builder/.arduino15/packages/arduino/tools/x86-linux-gcc/7.2.0 -prefs runtime.tools.serial-discovery.path=/home/builder/.arduino15/packages/builtin/tools/serial-discovery/1.1.0 -libraries /tmp/742609487/custom -libraries /tmp/742609487/pinned /tmp/742609487/basic

Multiple libraries were found for "TM1637.h"

Used: /home/builder/opt/libraries/latest/tm1637_driver_2_1_0

Not used: /home/builder/opt/libraries/latest/grove_4_digit_display_1_0_0

Using library tm1637_driver_2_1_0 at version 2.1.0 in folder: /home/builder/opt/libraries/latest/tm1637_driver_2_1_0

In file included from /home/builder/.arduino15/packages/arduino/hardware/megaavr/1.8.5/cores/arduino/Arduino.h:143:0,

from /tmp/742609487/build/sketch/basic.ino.cpp:1:

/home/builder/.arduino15/packages/arduino/hardware/megaavr/1.8.5/cores/arduino/USBCore.h:97:48: error: expected identifier before numeric constant

#define TEST_MODE 3

^

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:46:9: note: in expansion of macro 'TEST_MODE'

TEST_MODE = 0x48

^~~~~~~~~

/home/builder/.arduino15/packages/arduino/hardware/megaavr/1.8.5/cores/arduino/USBCore.h:97:48: error: expected '}' before numeric constant

#define TEST_MODE 3

^

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:46:9: note: in expansion of macro 'TEST_MODE'

TEST_MODE = 0x48

^~~~~~~~~

/home/builder/.arduino15/packages/arduino/hardware/megaavr/1.8.5/cores/arduino/USBCore.h:97:48: error: expected unqualified-id before numeric constant

#define TEST_MODE 3

^

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:46:9: note: in expansion of macro 'TEST_MODE'

TEST_MODE = 0x48

^~~~~~~~~

In file included from /home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/TM1637.h:6:0,

from /tmp/742609487/basic/basic.ino:72:

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:83:1: error: expected unqualified-id before 'public'

public:

^~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:86:5: error: expected unqualified-id before '{' token

{}

^

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:88:16: error: variable or field 'blink' declared void

void blink(Tasker::duration_type delay)

^~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:88:16: error: 'Tasker' has not been declared

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:103:18: error: variable or field 'fadeOut' declared void

void fadeOut(Tasker::duration_type delay)

^~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:103:18: error: 'Tasker' has not been declared

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:118:17: error: variable or field 'fadeIn' declared void

void fadeIn(Tasker::duration_type delay)

^~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:118:17: error: 'Tasker' has not been declared

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:135:21: error: variable or field 'scrollLeft' declared void

void scrollLeft(Tasker::duration_type delay)

^~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:135:21: error: 'Tasker' has not been declared

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:160:16: error: non-member function 'void off()' cannot have cv-qualifier

void off() const

^~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h: In function 'void off()':

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:162:9: error: 'sendToDisplay' was not declared in this scope

sendToDisplay<DataCommand_e::FIXED_ADDRESS, AddressCommand_e::C6H>(DisplayControl_e::DISPLAY_OFF);

^~~~~~~~~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:162:23: error: 'DataCommand_e' has not been declared

sendToDisplay<DataCommand_e::FIXED_ADDRESS, AddressCommand_e::C6H>(DisplayControl_e::DISPLAY_OFF);

^~~~~~~~~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:162:74: error: no match for 'operator>' (operand types are 'AddressCommand_e' and 'DisplayControl_e')

sendToDisplay<DataCommand_e::FIXED_ADDRESS, AddressCommand_e::C6H>(DisplayControl_e::DISPLAY_OFF);

~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:162:74: note: candidate: operator>(DisplayControl_e, DisplayControl_e) <built-in>

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:162:74: note: no known conversion for argument 1 from 'AddressCommand_e' to 'DisplayControl_e'

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:162:74: note: candidate: operator>(AddressCommand_e, AddressCommand_e) <built-in>

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:162:74: note: no known conversion for argument 2 from 'DisplayControl_e' to 'AddressCommand_e'

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h: At global scope:

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:165:42: error: non-member function 'void on(DisplayControl_e)' cannot have cv-qualifier

void on(DisplayControl_e brightness) const

^~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h: In function 'void on(DisplayControl_e)':

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:167:9: error: 'sendToDisplay' was not declared in this scope

sendToDisplay<DataCommand_e::FIXED_ADDRESS, AddressCommand_e::C6H>(brightness);

^~~~~~~~~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:167:23: error: 'DataCommand_e' has not been declared

sendToDisplay<DataCommand_e::FIXED_ADDRESS, AddressCommand_e::C6H>(brightness);

^~~~~~~~~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:167:74: error: no match for 'operator>' (operand types are 'AddressCommand_e' and 'DisplayControl_e')

sendToDisplay<DataCommand_e::FIXED_ADDRESS, AddressCommand_e::C6H>(brightness);

~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:167:74: note: candidate: operator>(DisplayControl_e, DisplayControl_e) <built-in>

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:167:74: note: no known conversion for argument 1 from 'AddressCommand_e' to 'DisplayControl_e'

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:167:74: note: candidate: operator>(AddressCommand_e, AddressCommand_e) <built-in>

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:167:74: note: no known conversion for argument 2 from 'DisplayControl_e' to 'AddressCommand_e'

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h: In function 'void reset(const String&)':

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:171:9: error: 'buffer_' was not declared in this scope

buffer_ = "";

^~~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:171:9: note: suggested alternative: 'epBuffer'

buffer_ = "";

^~~~~~~

epBuffer

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:173:22: error: 'toDisplayDigit' was not declared in this scope

auto d = toDisplayDigit(value[counter]);

^~~~~~~~~~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:173:22: note: suggested alternative: 'isDigit'

auto d = toDisplayDigit(value[counter]);

^~~~~~~~~~~~~~

isDigit

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:179:9: error: 'resetAnimation' was not declared in this scope

resetAnimation();

^~~~~~~~~~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:179:9: note: suggested alternative: 'Animation'

resetAnimation();

^~~~~~~~~~~~~~

Animation

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h: In function 'void clear()':

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:184:9: error: 'buffer_' was not declared in this scope

buffer_ = "";

^~~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:184:9: note: suggested alternative: 'epBuffer'

buffer_ = "";

^~~~~~~

epBuffer

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:185:42: error: 'totalDigits_' was not declared in this scope

for (size_t counter{}; counter < totalDigits_; ++counter)

^~~~~~~~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:187:9: error: 'refresh' was not declared in this scope

refresh();

^~~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h: In function 'void refresh()':

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:192:9: error: 'sendToDisplay' was not declared in this scope

sendToDisplay<DataCommand_e::AUTOMATIC_ADDRESS_ADDING, AddressCommand_e::C0H>(

^~~~~~~~~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:192:23: error: 'DataCommand_e' has not been declared

sendToDisplay<DataCommand_e::AUTOMATIC_ADDRESS_ADDING, AddressCommand_e::C0H>(

^~~~~~~~~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:193:17: error: 'brightness_' was not declared in this scope

brightness_,

^~~~~~~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:194:50: error: 'buffer_' was not declared in this scope

reinterpret_cast<const uint8_t*>(buffer_.c_str()),

^~~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:194:50: note: suggested alternative: 'epBuffer'

reinterpret_cast<const uint8_t*>(buffer_.c_str()),

^~~~~~~

epBuffer

In file included from /home/builder/.arduino15/packages/arduino/hardware/megaavr/1.8.5/cores/arduino/api/ArduinoAPI.h:52:0,

from /home/builder/.arduino15/packages/arduino/hardware/megaavr/1.8.5/cores/arduino/Arduino.h:23,

from /tmp/742609487/build/sketch/basic.ino.cpp:1:

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:195:41: error: 'totalDigits_' was not declared in this scope

static_cast<size_t>(min(totalDigits_, buffer_.length())));

^

/home/builder/.arduino15/packages/arduino/hardware/megaavr/1.8.5/cores/arduino/api/Common.h:41:19: note: in definition of macro 'min'

({ __typeof__ (a) _a = (a); \

^

In file included from /home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/TM1637.h:6:0,

from /tmp/742609487/basic/basic.ino:72:

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h: At global scope:

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:198:1: error: expected unqualified-id before 'private'

private:

^~~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h: In function 'void resetAnimation()':

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:212:9: error: 'currentAnimation' was not declared in this scope

currentAnimation = Animation::NONE;

^~~~~~~~~~~~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:212:9: note: suggested alternative: 'resetAnimation'

currentAnimation = Animation::NONE;

^~~~~~~~~~~~~~~~

resetAnimation

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h: At global scope:

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:235:15: error: 'DataCommand_e' has not been declared

template <DataCommand_e DATA_COMMAND, AddressCommand_e ADDRESS_COMMAND>

^~~~~~~~~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:236:63: error: non-member function 'void sendToDisplay(DisplayControl_e)' cannot have cv-qualifier

void sendToDisplay(DisplayControl_e displayControl) const noexcept

^~~~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h: In function 'void sendToDisplay(DisplayControl_e)':

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:238:9: error: 'mi2C_' was not declared in this scope

mi2C_.beginTransmission();

^~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:239:41: error: 'DATA_COMMAND' was not declared in this scope

mi2C_.send(static_cast<uint8_t>(DATA_COMMAND));

^~~~~~~~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:239:41: note: suggested alternative: 'ADC0_COMMAND'

mi2C_.send(static_cast<uint8_t>(DATA_COMMAND));

^~~~~~~~~~~~

ADC0_COMMAND

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h: At global scope:

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:249:15: error: 'DataCommand_e' has not been declared

template <DataCommand_e DATA_COMMAND, AddressCommand_e ADDRESS_COMMAND>

^~~~~~~~~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:250:99: error: non-member function 'void sendToDisplay(DisplayControl_e, const uint8_t*, size_t)' cannot have cv-qualifier

void sendToDisplay(DisplayControl_e displayControl, const uint8_t* values, size_t size) const noexcept

^~~~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h: In function 'void sendToDisplay(DisplayControl_e, const uint8_t*, size_t)':

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:252:9: error: 'mi2C_' was not declared in this scope

mi2C_.beginTransmission();

^~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:253:41: error: 'DATA_COMMAND' was not declared in this scope

mi2C_.send(static_cast<uint8_t>(DATA_COMMAND));

^~~~~~~~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:253:41: note: suggested alternative: 'ADC0_COMMAND'

mi2C_.send(static_cast<uint8_t>(DATA_COMMAND));

^~~~~~~~~~~~

ADC0_COMMAND

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:258:25: error: 'colon_' was not declared in this scope

mi2C_.send((colon_ || (dp_ & (1u << counter))) ? values[counter] | 0x80u : values[counter]);

^~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:258:25: note: suggested alternative: 'noTone'

mi2C_.send((colon_ || (dp_ & (1u << counter))) ? values[counter] | 0x80u : values[counter]);

^~~~~~

noTone

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:258:36: error: 'dp_' was not declared in this scope

mi2C_.send((colon_ || (dp_ & (1u << counter))) ? values[counter] | 0x80u : values[counter]);

^~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h: At global scope:

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:266:25: error: no matching function for call to 'MI2C::MI2C()'

MI2C mi2C_;

^~~~~

In file included from /home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:6:0,

from /home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/TM1637.h:6,

from /tmp/742609487/basic/basic.ino:72:

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/mI2C.h:12:5: note: candidate: MI2C::MI2C(uint8_t, uint8_t)

MI2C(uint8_t clkPin, uint8_t dataPin) noexcept : clkPin_{clkPin}, dataPin_{dataPin}

^~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/mI2C.h:12:5: note: candidate expects 2 arguments, 0 provided

In file included from /home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/TM1637.h:6:0,

from /tmp/742609487/basic/basic.ino:72:

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:271:5: error: 'Tasker' does not name a type

Tasker tasker;

^~~~~~

/home/builder/opt/libraries/latest/tm1637_driver_2_1_0/src/animator.h:289:1: error: expected declaration before '}' token

};

^

exit status 1

**Comment by AKJ7** [comment]
You certainly have 2 versions of the display installed. Uninstall both and reinstall mine.

**Comment by AKJ7** [comment]
Moreover, did you install this library from the Arduino IDE or did you just clone the repo?

**Comment by AKJ7** [comment]
I think the error is due to the fact that TEST_MODE which is a scope enum is already defined in the USBCore.h of the Nano. I will take care of it when i will get time.

**Comment by tdenslow** [comment]
Sorry for the delay getting back to you.  I knew I had tried a some older versions so multiple libraries doesn't surprise me.  I did get it directly from the Arduino IDE.  Sounds like you found the main issue.  Thanks!  

This really is a great library to work with.  Way easier than the TM1637Display library for a noob such as myself.  When I get a chance I'll shoot you some notes where I found the documentation might be a bit out of sync.
